Fix: lwc compiler errors on .jsx/.tsx files

SaurabhMulay999 opened the pull request:
## Details
Fix: LWC compiler errors on .jsx/.tsx files #4122. 
(Treating .jsx/.tsx  file extensions the same as .js & .ts.) 

## Does this pull request introduce a breaking change?
-   😮‍💨 No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?
-   🤞 No, it does not introduce an observable change.

nolanlawson commented:
Hi @SaurabhMulay999 thank you for the contribution! Could you please add a test to confirm the fix? The right place to add a test would be in [`transform-javascript.spec.ts`](https://github.com/salesforce/lwc/blob/32f4c5728a13c561c798d520c68ba3a0877043b7/packages/%40lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts). We just need to confirm that the new file extensions are handled correctly.

SaurabhMulay999 commented:
> Hi @SaurabhMulay999 thank you for the contribution! Could you please add a test to confirm the fix? The right place to add a test would be in [`transform-javascript.spec.ts`](https://github.com/salesforce/lwc/blob/32f4c5728a13c561c798d520c68ba3a0877043b7/packages/%40lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts). We just need to confirm that the new file extensions are handled correctly.

Sure, I'll try.

SaurabhMulay999 commented:
Hi @nolanlawson ,
I have tried writing a unit test. 

Code:

describe('transformFile function',()=>{
    it('should be tranform Typescript React file (.tsx)',()=>{
        const src=`import React from 'react'; \n const App=()=><div>Hello World</div>;`
        const fileName='app.tsx';
        const options={namespace:'c',name:'app'};
        const result=transformSync(src,fileName,options);
        expect(result.code).toBeDefined();
        expect(result.map).toBeDefined();
        expect(result.warnings).toBeUndefined();
        expect(result.cssScopeTokens).toBeUndefined();

    });

    it('should be tranform Typescript React file (.jsx)',()=>{
        const src=`import React from 'react'; \n const App=()=><div>Hello World</div>;`
        const fileName='app.jsx';

        const options={namespace:'c',name:'app'};
        const result=transformSync(src,fileName,options);
        expect(result.code).toBeDefined();
        expect(result.map).toBeDefined();
        expect(result.warnings).toBeUndefined();
        expect(result.cssScopeTokens).toBeUndefined();

    });
});

after debug it was failing saying that. 

Error:
Support for the experimental syntax 'jsx' isn't currently enabled.qq

I have checked some online resources it was suggesting to install @babel/preset-react.

nolanlawson commented:
@SaurabhMulay999 It should be fine to just put actual JavaScript in there rather than JSX.

We do not need to test transpiling JSX into JavaScript; just that the `.jsx`/`.tsx` extensions are supported.

SaurabhMulay999 commented:
> @SaurabhMulay999 It should be fine to just put actual JavaScript in there rather than JSX.
> 
> We do not need to test transpiling JSX into JavaScript; just that the `.jsx`/`.tsx` extensions are supported.
> 
> In a real app setup (e.g. using Rollup), the consumer of LWC would be responsible for ensuring that there are plugins in place to do the transformation.

Okay. I've added the JS format. I have ran it locally it was successful. Let me know if any modifications are required.

nolanlawson commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
+
+describe('transformFile function', () => {
+    it('should be tranform Typescript file with ext (.tsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+    `;
```
```suggestion
        const src = `
            import { LightningElement } from 'lwc';
            export default class Foo extends LightningElement {}
        `;
```

Just slightly nicer formatting.

nolanlawson commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
+    });
+
+    it('should be tranform Javascript file with ext (.jsx)', () => {
+        const src = `
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+        `;
```
```suggestion
        const src = `
            import { LightningElement } from 'lwc';
            export default class Foo extends LightningElement {}
        `;
```

Ditto.

nolanlawson commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+    `;
+        const fileName = 'app.ts';
```
```suggestion
        const fileName = 'app.tsx';
```

nolanlawson commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
+        import { LightningElement } from 'lwc';
+        export default class Foo extends LightningElement {}
+        `;
+        const fileName = 'app.js';
```
```suggestion
        const fileName = 'app.jsx';
```

SaurabhMulay999 commented:
> @SaurabhMulay999 Could you please sign the CLA as described above?

Hey @nolanlawson,
I don't know what is the issue but previously i already have signed the cla. that site redirect me and shows that, that i already have signed the cla. but this time system failed to verify my email. is there any workaround? like to resign it.

wjhsf commented:
> Commit authors must be associated with GitHub users

I think you'll have to redo the commits using the email address associated with your GitHub account. (Or add the email address from your commits to your GitHub account, but the one used [here](https://github.com/salesforce/lwc/commit/c222bd30c398359ede528aa41d25a0d872aa70a1.patch) doesn't seem to be a real email address, so I don't think GitHub will let you.)

ekashida reviewed (commented):
It would be nice to loop over an array of all the supported file extensions. Something like:
```js

    describe('file extension support', () => {
        function testFileExtensionSupport(ext: string) {
            it(`should support ${ext} file extension`, () => {
                const src = `
                    import { LightningElement } from 'lwc';
                    export default class Foo extends LightningElement {}
                `;
                const options = { namespace: 'c', name: 'foo' };
                const result = transformSync(src, `foo${ext}`, options);
                expect(result.code).toBeDefined();
                expect(result.map).toBeDefined();
                expect(result.warnings).toBeUndefined();
                expect(result.cssScopeTokens).toBeUndefined();
            });
        }
        ['.js', '.jsx', '.ts', '.tsx'].forEach(testFileExtensionSupport);
    });
```

SaurabhMulay999 commented:
> > Commit authors must be associated with GitHub users
> 
> I think you'll have to redo the commits using the email address associated with your GitHub account. (Or add the email address from your commits to your GitHub account, but the one used [here](https://github.com/salesforce/lwc/commit/c222bd30c398359ede528aa41d25a0d872aa70a1.patch) doesn't seem to be a real email address, so I don't think GitHub will let you.)

I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.

SaurabhMulay999 commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,34 @@ describe('sourcemaps', () => {
     });
 });
+
+describe('transformFile function', () => {
```
I think it's better as you have suggested above to loop over all the extensions.

wjhsf commented:
> I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.
If you use your salesforce email then you don't need to sign the CLA, because it will recognize you as an internal user.

SaurabhMulay999 commented:
> > I think it tried to fetch my official salesforce email. I'll try adding the personal email and then recommit.
> > If you use your salesforce email then you don't need to sign the CLA, because it will recognize you as an internal user.

Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.

ekashida reviewed (approved):
lgtm!

SaurabhMulay999 commented:
> > Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.
> 
> The older commits still show the invalid email address. I believe you will need to do an [interactive rebase](https://www.sitepoint.com/git-interactive-rebase-guide/) or squash-merge followed by a force-push to fix this.

Okay. I'll try interactive rebase.

SaurabhMulay999 commented:
> > Yeah, but this is my personal account. I have added the Name as well email in the latest commit. It should have now recognise me as a GitHub user.
> 
> The older commits still show the invalid email address. I believe you will need to do an [interactive rebase](https://www.sitepoint.com/git-interactive-rebase-guide/) or squash-merge followed by a force-push to fix this.

Hi @nolanlawson , 
I have changed the commit author for all the commits. and also have squashed (2 to N) commits together. Thanks for the suggestion, I have learned a lot.

nolanlawson commented on packages/@lwc/compiler/src/transformers/__tests__/transform-javascript.spec.ts:
```diff
@@ -269,3 +269,21 @@ describe('sourcemaps', () => {
+            expect(result.cssScopeTokens).toBeUndefined();
+        });
+    }
+    ['.js', '.jsx', '.ts', '.tsx'].forEach(testFileExtensionSupport);
```
Nice!

nolanlawson commented:
/nucleus test
